compiler/src/dotty/tools/dotc/core/ConstraintHandling.scala

```diff
@@ -354,12 +358,13 @@ trait ConstraintHandling {
       case WildcardType(optBounds) => optBounds.exists && isSingleton(optBounds.bounds.hi)
       case _ => isSubTypeWhenFrozen(tp, defn.SingletonType)
 
-    def isEnumBound(tp: Type): Boolean =
-      bound.isValueType && bound.typeSymbol.is(Enum, butNot=JavaDefined) && tp <:< bound
+    def isEnum(tp: Type): Boolean = tp match
+      case WildcardType(optBounds) => optBounds.exists && isEnum(optBounds.bounds.hi)
+      case _ => tp.typeSymbol.is(Enum, butNot=JavaDefined)
 
     val wideInst =
-      if isSingleton(bound) || isEnumBound(bound) then inst
-      else dropSuperTraits(widenOr(widenSingle(inst)))
+      if isSingleton(bound) || isEnum(bound) then inst
+      else dropSuperTraits(widenOr(widenEnum(widenSingle(inst))))
```

Comment:
I think that's still not quite right: if the upper bound is an enum, a singleont should still be widened unless it's the type of an enum case.

Reply:
do you mean remove singletons except the term ref of a singleton enum case?

Reply:
yeah

Reply:
this seems to get a lot trickier when unions of singletons are involved

Reply:
new steps are `widenSingle` > `widenOr` > `[widenEnumCase]` > `dropSuperTraits`, where `widenOr` is intercepted so that singletons of module or enum value do not widen